ftw/upgrade/command/jsonapi.py

```diff
@@ -0,0 +1,206 @@
+        self.session.auth = (username, password)
+        self.site = site
+
+    def GET(self, action, site=None, **kwargs):
+        return self._make_request('GET', action, site=site, **kwargs)
+
+    def POST(self, action, site=None, **kwargs):
+        return self._make_request('POST', action, site=site, **kwargs)
+
+    def _make_request(self, method, action, site=None, **kwargs):
+        url = get_api_url(action, site=site or self.site)
+        response = self.session.request(method.upper(), url, **kwargs)
+        response.raise_for_status()
+        return response
+
+
+def add_requestor_authentication_argument(argparse_command):
+    argparse_command.add_argument(
+        '--auth',
+        help='Authentication information: "<username>:<password>"')
+
+
+def add_site_path_argument(argparse_command):
+    group = argparse_command.add_mutually_exclusive_group(required=True)
+    group.add_argument(
+        '--site', '-s',
+        help='Path to the Plone site.')
+    group.add_argument(
+        '--pick-site', '-S', action='store_true',
+        help='Automatically pick the Plone site, when there is exactly one.')
+    group.add_argument(
+        '--last-site', '-L', action='store_true',
+        help='Automatically pick the last Plone site ordered by path.')
+
+
+def add_json_argument(argparse_command):
+    argparse_command.add_argument('--json',
+                                  action='store_true',
+                                  help='Print result as JSON.')
+
+
+def with_api_requestor(func):
+    def func_wrapper(args):
+        default_auth = os.environ.get('UPGRADE_AUTHENTICATION', None)
+        auth_value = args.auth or default_auth
+        if not auth_value:
+            print 'ERROR: No authentication information provided.'
+            print 'Use either the --auth param or the UPGRADE_AUTHENTICATION' + \
+                ' environment variable for providing authentication information' + \
+                ' in the form "<username>:<password>".'
+            sys.exit(1)
+
+        if len(auth_value.split(':')) != 2:
+            print 'ERROR: Invalid authentication information "{0}".'.format(
+                auth_value)
+            print 'A string of form "<username>:<password>" is required.'
+            sys.exit(1)
+
+        site = get_plone_site_by_args(args, auth_value)
+        requestor = APIRequestor(*auth_value.split(':'), site=site)
+        return func(args, requestor)
+    func_wrapper.__name__ = func.__name__
+    func_wrapper.__doc__ = func.__doc__
+    return func_wrapper
+
+
+def error_handling(func):
+    def func_wrapper(*args, **kwargs):
+        try:
+            return func(*args, **kwargs)
+        except NoRunningInstanceFound:
+            print 'ERROR: No running Plone instance detected.'
+            sys.exit(1)
+        except HTTPError, exc:
+            try:
+                response = exc.response
+                mimetype = (response.headers.get('Content-Type')
```

Comment:
A slightly nicer way of doing this would be `mimetype = cgi.parse_header(content_type)[0]` (that's [how `requests` does it](https://github.com/kennethreitz/requests/blob/master/requests/utils.py#L319)).